Add unit test for calculating min/max elevation

flother opened the pull request:
This is a pull request on top of PR #1300. It adds a unit test for that PR's bug fix to [`Transform.coveringTiles()`](https://github.com/maplibre/maplibre-gl-js/blob/577f200072b9a55ce9b841bc04a21cde7603e01c/src/geo/transform.ts#L334). The affected code has been pulled out into a separate method, `Transform.getTileMinMaxElevation()`, so that it can be tested in isolation. (That method should probably be on [`DEMData`](https://github.com/maplibre/maplibre-gl-js/blob/main/src/data/dem_data.ts), but time has run away from me.)

This test fails on `main` but passes on the `covering_tiles_3d` branch.

prozessor13 commented:
Hallo Matt,

thx for your pull-request, but i am not shure if transform.ts ist the correct place to implement `getTileMinMaxElevation`. If this method is needed, i think it would be better to put this into terrain.ts `getMinMaxElevation(tileID)` or something like that.

Regards. Max.

HarelM commented:
It might be even better to move the entire content of the `if (options.terrain)...` to `terrain.ts` file so that everything related to terrain is there and should be tested separately?
I generally agree with @prozessor13 comment I think...
Thanks @flother for pushing this forward!

flother commented:
Thanks for the suggestions — I knew it felt wrong to have the `getTileMinMaxElevation` method in `transform.ts`. The `Terrain` class is a much better fit. I've added commits that move the code.

> If this method is needed, i think it would be better to put this into terrain.ts `getMinMaxElevation(tileID)` or something like that.

I've moved it to `Terrain.getMinMaxElevation(tile)` (`tile` instead of `tileID`) to avoid having to instantiate a new `Tile` within `getMinMaxElevation`. That makes it easier to test in isolation.

> It might be even better to move the entire content of the `if (options.terrain)...` to `terrain.ts` file so that everything related to terrain is there and should be tested separately?

I've moved some, but not all, of the code in the `if` statement into `getMinMaxElevation`. Moving it all would mean the method would need lots of variables passed as arguments — `childZ`, `it.wrap`, `childZ`, `childX`, `childY`, `this.elevation` — and that felt like overkill. But I'm happy to do that if you'd prefer it that way.

HarelM commented:
I think moving the `getTerrainData` and passing only tileID would be just a but more clean, but otherwise this PR looks good.
I can't help but ask for a render test to make sure we the entire viewport is not white as can be seen in the relevant issue, I understand if this is a big request... :-)

flother commented:
I've changed `Terrain.getMinMaxElevation()` in the latest commit so it takes a `tileID` argument instead of `tile`.

The render tests are new to me, I'd need some help creating one of those. I couldn't see any obvious terrain render tests in the [`./test/integration/render`](https://github.com/maplibre/maplibre-gl-js/tree/main/test/integration/render) directory already. Is there something I could look at for inspiration?

HarelM commented:
Check out the readme of how to write the render tests:
https://github.com/maplibre/maplibre-gl-js/tree/main/test/integration#writing-new-tests
Terrain 3D doesn't have a render test yet unfortunately, but this is why it's even more important :-)
You basically need terrain tiles, which exits in this repo for the hillshading render tests, a style.json file and the rest is done by the code.
I would consider trying to mesh a hillshading and raster test to get the right results....

HarelM commented:
I'll merge this into the relevant branch as I think this is a good improvement.
We can decide if we want to merge the other branch before or after the render test is added.
